### PR TITLE
Initial support for new WebAssembly JS API

### LINKED ISSUE #1882: WASM - Copy binary buffer
We should make a copy of the binary buffer and use this for compilation instead of the source.
The reason is the source can be modified between the time it is passed and when we do the actual compilation.

### LINKED ISSUE #1884: WASM - Async Javascript API
Implement the final javascript API for the MVP

### PR DESCRIPTION
**Pull request by MikeHolman**
Adds support for the following:
WebAssembly.compile
WebAssembly.validate
new WebAssembly.Module
new WebAssembly.Instance
new WebAssembly.RuntimeError
new WebAssembly.CompileError
new WebAssembly.Memory
WebAssembly.Memory.prototype.grow
WebAssembly.Memory.prototype.buffer

Still remaining is:
WebAssembly.Table

Closes #1884 
Closes #1882

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/1821)

### CODE REVIEW

**Comment by MikeHolman** [comment]
@Cellule There are failures with baselines since I am now using WebAssembly.CompileError instead of having RuntimeError. I assume that these will have to be changed again once int64 is in, so I'll wait on that to do the baseline update.

**Comment by MikeHolman** [comment]
@boingoing Can you also take a look at JavascriptLibrary changes? I want to make sure I'm doing correct thing to add new types and objects.

File: Build/Chakra.Core.sln [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 14
-VisualStudioVersion = 14.0.25123.0
+VisualStudioVersion = 14.0.25420.1
```
Comment (dilijev) [inline Build/Chakra.Core.sln]: Any reason this file needs to be updated?

File: lib/Runtime/Base/JnDirectFields.h [hunk 2]
```diff
@@ -275,9 +275,19 @@ ENTRY(Map)
 ENTRY(Math)
 
 // Wasm
+// TODO: remove old Wasm object
```
Comment (dilijev) [inline lib/Runtime/Base/JnDirectFields.h]: Is there a work item tracking this?

Comment (MikeHolman) [inline lib/Runtime/Base/JnDirectFields.h]: Opened https://github.com/Microsoft/ChakraCore/issues/1834 to track this

File: Build/Chakra.Core.sln (hunk 1 above)
Comment (MikeHolman) [inline Build/Chakra.Core.sln]: IDK, VS wanted to update it. I didn't see any reason to tell VS otherwise.

**Comment by Cellule** [comment]
Could you add a more complex test mostly for WebAssembly.Memory

---

Reviewed 22 of 44 files at r1, 48 of 48 files at r2.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV7vfw8zzQC5xnpC5DM:-KV7vfw8zzQC5xnpC5DN:bsd089n) (no related file):*
Since you removed the logic to handle LazyTraps when deferparse is not enable, could you remove that flag from configflagslist.h

---

*[lib/Runtime/Language/InterpreterStackFrame.cpp, line 2980 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV80wKElIUf55v2N7Nz:-KV80wKElIUf55v2N7O-:b5pehzm) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Language/InterpreterStackFrame.cpp#L2980)):*
> ```cpp
>         m_localSlots[AsmJsFunctionMemory::ModuleEnvRegister] = frame->GetItem(0);
> #ifdef ENABLE_WASM
>         if (func->GetFunctionBody()->IsWasmFunction())
> ```

You added the same code in AsmJsModule.cpp, coudl we make that a function in ScriptFunction ?

---

*[lib/Runtime/Language/InterpreterStackFrame.h, line 366 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV81FRo9_M6RuUs0Bxb:-KV81FRo9_M6RuUs0Bxc:by7gxlf) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Language/InterpreterStackFrame.h#L366)):*
> ```objc
>         Var DebugProcess();
> 
>         ArrayBuffer * LoadWAsmJsArrayBuffer() const;
> ```

You haven't implemented that method, I guess this is related to my comment in the .cpp file

---

*[lib/Runtime/Library/ArrayBuffer.h, line 132 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV81SaiZC0fbsmpe9Pu:-KV81SaiZC0fbsmpe9Pv:b-8y0f5r) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/ArrayBuffer.h#L132)):*
> ```objc
>         virtual bool IsArrayBuffer() override { return true; }
>         virtual bool IsSharedArrayBuffer() override { return false; }
>         virtual ArrayBuffer * GetAsArrayBuffer() override { return ArrayBuffer::FromVar(this); }
> ```

I think we should ask someone who knows more about the ArrayBuffer to see if it's okay to expose those methods or if there is another api we should use

---

*[lib/Runtime/Library/Chakra.Runtime.Library.vcxproj.filters, line 95 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV7m6Ok1szskToqfthn:-KV7m6Ok1szskToqftho:bgaqskv) ([raw file](https://github.com/microsoft/chakracore/blob/0f1795e21f196aa4ee607d69ed36bf9b6a62b323/lib/Runtime/Library/Chakra.Runtime.Library.vcxproj.filters#L95)):*
> ```XML
>     <ClCompile Include="$(MSBuildThisFileDirectory)JavascriptStringObject.cpp" />
>     <ClCompile Include="$(MSBuildThisFileDirectory)MathLibrary.cpp" />
>     <ClCompile Include="$(MSBuildThisFileDirectory)NullEnumerator.cpp" />
> ```

This file was just missing ?

---

*[lib/Runtime/Library/Chakra.Runtime.Library.vcxproj.filters, line 115 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV7mAvKsLGIYRCmVNCS:-KV7mAvKsLGIYRCmVNCT:bmcyv77) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/Chakra.Runtime.Library.vcxproj.filters#L115)):*
> $(MSBuildThisFileDirectory)

Missing

---

*[lib/Runtime/Library/JavascriptFunction.cpp, line 2100 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV82DPik5LG58dg3Vp4:-KV82DPik5LG58dg3Vp5:b-hx27p6) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/JavascriptFunction.cpp#L2100)):*
> ```cpp
>         if (isAsmJs)
>         {
>             Assert(!funcBody->IsWasmFunction());
> ```

Did we not implement the big array for wasm ?

---

*[lib/Runtime/Library/JavascriptLibrary.cpp, line 288 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV82m7sHZagrsnYZ1gj:-KV82m7t7h5ka9aCbUDq:b-n76a1) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/JavascriptLibrary.cpp#L288)):*
> ```cpp
> 
> #define INIT_ERROR_PROTO(field, initFunc) \
>         if (scriptContext->GetConfig()->IsES6PrototypeChain()) \
> ```

What happens with WebAssembly errors if this is not enabled ?

---

*[lib/Runtime/Library/JavascriptLibraryBase.h, line 173 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV84lOHezM3QqD5bql_:-KV84lOHezM3QqD5bqla:b-qtlqas) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/JavascriptLibraryBase.h#L173)):*
> ```objc
>         RuntimeFunction* asyncFunctionConstructor;
> 
>         RuntimeFunction* webAssemblyModuleConstructor;
> ```

Delete this please

---

*[lib/Runtime/Library/WebAssembly.cpp, line 27 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV85Vtsm75_6nd9EYHz:-KV85Vtsm75_6nd9EYI-:bo9vfxz) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/WebAssembly.cpp#L27)):*
> ```cpp
>     try
>     {
>         if (args.Info.Count < 2)
> ```

This part is almost exactly the same as WebAssemblyModule::NewInstance can we combine those ?

---

*[lib/Runtime/Library/WebAssemblyInstance.cpp, line 59 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV7syzYkSdpgHHk3QGP:-KV7syzZcUHqFBAQutx5:bctkqgi) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/WebAssemblyInstance.cpp#L59)):*
> ```cpp
>     if (args.Info.Count < 2 || !WebAssemblyModule::Is(args[1]))
>     {
>         JavascriptError::ThrowTypeError(scriptContext, WASMERR_NeedModule, _u("WebAssembly.Instance"));
> ```

Shouldn't the last param be the name of that argument, ie: `moduleObject`

---

*[lib/Runtime/Library/WebAssemblyInstance.cpp, line 72 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV7tDJHQQxdSIdX_8Rj:-KV7tDJHQQxdSIdX_8Rk:bev3wyh) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/WebAssemblyInstance.cpp#L72)):*
> ```cpp
>         if (!JavascriptOperators::IsUndefined(args[2]) && !JavascriptOperators::IsObject(args[2]))
>         {
>             JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedObject, _u("WebAssembly.Instance"));
> ```

Shouldn't the last param be the name of that argument, ie: `importObject`

---

*[lib/Runtime/Library/WebAssemblyInstance.cpp, line 79 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV7tEyf8u487s9_AO79:-KV7tEyf8u487s9_AO7A:b-ya81ng) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/WebAssemblyInstance.cpp#L79)):*
> ```cpp
>     if (module->GetImportCount() > 0 && !JavascriptOperators::IsObject(importObject))
>     {
>         JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedObject, _u("WebAssembly.Instance"));
> ```

Same as above

---

*[lib/Runtime/Library/WebAssemblyInstance.cpp, line 135 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV7tVGfeAPxFA-CDjbG:-KV7tVGfeAPxFA-CDjbH:b-vqnpow) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/WebAssemblyInstance.cpp#L135)):*
> ```cpp
>         localModuleFunctions[i] = funcObj;
> 
>         FunctionBody* body = funcObj->GetFunctionBody();
> ```

If we call `new WebAssembly.Instance(mod1)` twice, we will overwrite the entrypoint to something possibly wrong.
We need to determine if that function body was already used by another Instance and not update the entry points.

eg:
```js
WebAssembly.compile(buf).then(function(val)
{
  const inst1 = new WebAssembly.Instance(val, {});
  inst1.exports.foo();
  const inst2 = new WebAssembly.Instance(val, {});
  inst2.exports.foo();
})
```
We would set foo to deferparse entry point, then call it, then set it again to deferparse entry point and then the WasmReaderInfo will already be null.

Maybe we can use WasmReaderInfo to determine if we need to set the entry point.

---

*[lib/Runtime/Library/WebAssemblyInstance.cpp, line 177 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV7zCMftl0c1YEuYK-q:-KV7zCMgtbT2gJ3R7V9c:b-fv75jv) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/WebAssemblyInstance.cpp#L177)):*
> ```cpp
>             const uint32 offset = segment->getDestAddr(wasmModule);
>             const uint32 size = segment->getSourceSize();
>             if (UInt32Math::Add(offset, size) > buffer->GetByteLength())
> ```

Move this check inside `size > 0`

---

*[lib/Runtime/Library/WebAssemblyModule.h, line 34 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV7kTOlIiFZklFg9lI7:-KV7kTOlIiFZklFg9lI8:burosbb) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/WebAssemblyModule.h#L34)):*
> ```objc
> }
> 
> // TODO: merge with WasmModule
> ```

Outdated todo since WasmModule doesn't exist anymore

---

*[lib/Runtime/Types/EdgeJavascriptTypeId.h, line 115 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV7k1f5n1NdSK-GmwI6:-KV7k1f5n1NdSK-GmwI7:b-synk3a) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Types/EdgeJavascriptTypeId.h#L115)):*
> ```objc
>     TypeIds_SharedArrayBuffer = 70,
> 
>     TypeIds_WebAssemblyModule = 71,
> ```

Doesn't these values conflict with what's following ?

---

*[test/wasm/basic.js, line 6 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV7hkBNTJMh4jCOmuPH:-KV7hkBOTQGz_oxA7SmI:bhht7wi) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/test/wasm/basic.js#L6)):*
> ```JavaScript
> //-------------------------------------------------------------------------------------------------------
> 
> var module = WebAssembly.compile(readbuffer('basic.wasm'));
> ```

nit: this is technically not the `module` it's a promise.
You can rewrite this
```js
WebAssembly.compile(readbuffer('basic.wasm')).then(module => {
  const instance = new WebAssembly.Instance(module, {test: {foo: function(a){print(a); return 2;}}});
  ...
```

---

*[test/WasmSpec/baselines/memory.baseline, line 1 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV7iEFFhQbCe9qDUf44:-KV7iEFGmrwT38QE1tH9:b-8mcpqq) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/test/WasmSpec/baselines/memory.baseline#L1)):*
> ```
> Unexpected Error while compiling memory.17.wasm
> ```

Why is there a new error in this file ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-:-KV7jTvUpwzgtGcE2U3c:b-ygvpdm)*

**Comment by Cellule** [comment]
Reviewed 9 of 10 files at r3.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1821)*

**Comment by Cellule** [comment]
Reviewed 1 of 10 files at r3.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1821)*

**Comment by MikeHolman** [comment]
*[lib/Runtime/Library/ArrayBuffer.h, line 132 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-KV81SaiZC0fbsmpe9Pu:-KVaYYtHt8g3EnhWZdCg:bowwnji) ([raw file](https://github.com/microsoft/chakracore/blob/5736e92fbb40c25c0233e0a3224a11ebc4d26499/lib/Runtime/Library/ArrayBuffer.h#L132)):*
<details><summary><i>Previously, Cellule (Michael Ferris) wrote…</i></summary>
> I think we should ask someone who knows more about the ArrayBuffer to see if it's okay to expose those methods or if there is another api we should use
</details>
I was one who wrote this method. Should be ok to expose.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1821)*

**Comment by Cellule** [comment]
Reviewed 23 of 23 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1821)*

**Comment by Cellule** [comment]
There are still 4 comments that haven't been addressed on Reviewable

**Comment by Cellule** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1821#-:-KVavr6KWyRgiuQ8WYw0:bnfp4nl)*

**Pull request merged**